File: backend/driver/src/main/java/dev/tpcoder/empfriendly/driver/ExceptionHandler.java
package dev.tpcoder.empfriendly.driver;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.tpcoder.empfriendly.driver.model.ErrorResponse;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import org.springframework.core.annotation.Order;
import org.springframework.core.io.buffer.DataBuffer;
import org.springframework.core.io.buffer.DataBufferUtils;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebExceptionHandler;
import reactor.core.publisher.Mono;

@Component
@Order(-2)
public class ExceptionHandler implements WebExceptionHandler {

  private final ObjectMapper objectMapper;

  public ExceptionHandler(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  @Override
  public Mono<Void> handle(ServerWebExchange exchange, Throwable ex) {
    return Mono.defer(() -> {
      try {
        final ServerHttpResponse response = exchange.getResponse();
        final HttpHeaders headers = response.getHeaders();
        final ErrorResponse error = ErrorResponse.serverError(ex.getMessage());

        response.setStatusCode(HttpStatus.valueOf(error.errorStatus()));
        headers.put(HttpHeaders.CONTENT_TYPE, Collections.singletonList(MediaType.APPLICATION_JSON_VALUE));

        final String json = objectMapper.writeValueAsString(error);
        final DataBuffer buffer = response.bufferFactory()
            .wrap(json.getBytes(StandardCharsets.UTF_8));

        return response.writeWith(Mono.just(buffer))
            .doOnError(e -> DataBufferUtils.release(buffer));
      } catch (final Exception e) {
        return Mono.error(e);
      }
    });
  }
}

File: backend/general/src/main/kotlin/dev/tpcoder/empfriendly/general/config/ExceptionHandler.kt
package dev.tpcoder.empfriendly.general.config

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.core.annotation.Order
import org.springframework.core.io.buffer.DataBufferUtils
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebExceptionHandler
import reactor.core.publisher.Mono
import java.nio.charset.StandardCharsets

@Component
@Order(-2)
class ExceptionHandler(private val objectMapper: ObjectMapper) : WebExceptionHandler {
    override fun handle(exchange: ServerWebExchange, ex: Throwable): Mono<Void> {
        return Mono.defer {
            try {
                val response =
                    exchange.response
                val headers = response.headers
                val error: ErrorResponse = ErrorResponse.serverError(ex.message)
                response.statusCode = HttpStatus.valueOf(error.errorStatus())
                headers[HttpHeaders.CONTENT_TYPE] = listOf(MediaType.APPLICATION_JSON_VALUE)
                val json = objectMapper.writeValueAsString(error)
                val buffer = response.bufferFactory()
                    .wrap(json.toByteArray(StandardCharsets.UTF_8))
                val monoBuffer = Mono.just(buffer)
                return@defer response.writeWith(monoBuffer)
                    .doOnError { DataBufferUtils.release(buffer) }
            } catch (e: Exception) {
                return@defer Mono.error<Void>(e)
            }
        }
    }
}